### DISCUSSION TITLE
Curvature calculation for surfacetension force.

### DISCUSSION THREAD

**Discussion by MashrurShejan** [idaholab/moose#26717] [Q&A Modules: Navier-Stokes]
Hello, I am trying to incorporate curvature calculation in My surface tension force kernel which is giving me error currently. Here is the kernel code below-
```
#include "SurfaceTensionForce.h"

registerMooseObject("NavierStokesApp", SurfaceTensionForce);

InputParameters SurfaceTensionForce::validParams()
{
  InputParameters params = ADKernel::validParams();
  params.addClassDescription("ADKernel for computing surface tension force");
  params.addRequiredParam<MaterialPropertyName>("sigma", "The name of the surface tension coefficient material property.");
  params.addRequiredCoupledVar("alpha", "The name of the volume fraction variable.");
  return params;
}

SurfaceTensionForce::SurfaceTensionForce(const InputParameters & parameters) :
  ADKernel(parameters),
  _sigma(getADMaterialProperty<Real>("sigma")),
  _grad_alpha(adCoupledGradient("alpha"))
{
}

ADReal SurfaceTensionForce::computeQpResidual()
{
  ADReal kappa = computeKappa();
  ADReal sigma_interpolated = _sigma[_qp];
  return sigma_interpolated * kappa * _grad_alpha[_qp] * _grad_test[_i][_qp];
}

ADReal SurfaceTensionForce::computeKappa() const
{
  ADReal grad_alpha_mag = _grad_alpha[_qp].norm();

  if (grad_alpha_mag > 1e-6) //To avoid deviding by zero
  {
    ADRealVectorValue norm_alpha = _grad_alpha[_qp] / grad_alpha_mag;

    
    ADReal kappa = 0.0;
    for (unsigned int i = 0; i < LIBMESH_DIM; ++i)
    {
      // Placeholder 
      //kappa += _grad_norm_alpha[_qp][i]
    }

    return kappa;
  }
  else
  {
    return 0.0;
  }
}
```
The code compiles fine when the the kappa calculation is commented out. I am not sure what I am doing wrong. Can anyone review the code and let me know the issue? Please note that I am not very proficient with C++ or MOOSE Library.

**Comment by GiudGiud** [comment]
Hello

If looks like it should not be grad_norm_alpha but norm_alpha

The variables don't have the right name. 

Also to access a component of a vector you use (i) not [i]

**Reply by MashrurShejan** [reply]
Hi! According to the curvature formula I need to calculate the divergence of the normalized vector alpha denoted by _grad_alpha[_qp] / grad_alpha_mag. I am not sure if kappa += norm_alpha[_qp] (i) is what it is supposed to be.

**Reply by GiudGiud** [reply]
Hello

sure but you did not define ` _grad_norm_alpha` anywhere before using it below
```
      //kappa += _grad_norm_alpha[_qp][i]
```

You need to define it first.

**Reply by MashrurShejan** [reply]
I was trying to calculate the gradient of norm alpha using _grad_norm_alpha and summing them together for different dimensions using  kappa+=_grad_norm_alpha(i) to calculate the divergence here. But I guess this is not how it will work. In that case,  what should be the expression to define the divergence of  norm_alpha here?
